tests/fs/api.public.test.ts

```diff
@@ -0,0 +1,292 @@
+
+          await fs.write(fromPath, val)
+          await fs.mv(fromPath, toPath)
+
+          const file = await fs.read(toPath)
+          const decodedContent = decode(file, type)
+
+          expect(decodedContent).toEqual(val)
+        })
+    )
+  })
+
+  it('makes directories', async () => {
+    const fs = await emptyFilesystem()
+
+    await fc.assert(
+      fc.asyncProperty(
+        pathSegment(), async data => {
+          const dirpath = path.directory('public', data[0])
+
+          await fs.mkdir(dirpath)
+
+          expect(await fs.exists(dirpath)).toEqual(true)
+        }),
+      { numRuns: 100 }
+    )
+  })
+
+  it('removes directories it makes', async () => {
+    const fs = await emptyFilesystem()
+
+    await fc.assert(
+      fc.asyncProperty(
+        pathSegment(), async data => {
+          const dirpath = path.directory('public', data[0])
+
+          await fs.mkdir(dirpath)
+          await fs.rm(dirpath)
+
+          expect(await fs.exists(dirpath)).toEqual(false)
+        })
+    )
+  })
+
+  it('writes files to a directory', async () => {
+    const fs = await emptyFilesystem()
+    const dirpath = path.directory('public', 'testDir')
+    await fs.mkdir(dirpath)
+
+    await fc.assert(
+      fc.asyncProperty(
+        fc.tuple(pathSegment(), fileContent()), async data => {
+          const filepath = path.file('public', 'testDir', data[0])
+          const [val] = data[1]
+
+          await fs.write(filepath, val)
+
+          expect(await fs.exists(filepath)).toEqual(true)
+        })
+    )
+  })
+
+  it('lists files written to a directory', async () => {
+    const fs = await emptyFilesystem()
+    const dirpath = path.directory('public', 'testDir')
+    await fs.mkdir(dirpath)
+
+    await fc.assert(
+      fc.asyncProperty(
+        fc.tuple(pathSegment(), fileContent()), async data => {
+          const filepath = path.file('public', 'testDir', data[0])
+          const [val] = data[1]
+
+          await fs.write(filepath, val)
+          const listing = await fs.ls(dirpath)
+
+          expect(data[0] in listing).toEqual(true)
+        })
+    )
+  })
+
+  it('moves files into a directory', async () => {
+    const fs = await emptyFilesystem()
+    const dirpath = path.directory('public', 'testDir')
+    await fs.mkdir(dirpath)
+
+    await await fc.assert(
+      fc.asyncProperty(
+        fc.tuple(pathSegmentPair(), fileContent()), async data => {
+          const [from, to] = data[0]
+          const [val] = data[1]
+          const fromPath = path.file('public', from)
+          const toPath = path.file('public', to)
+
+          await fs.write(fromPath, val)
+          await fs.mv(fromPath, toPath)
+          const fromExists = await fs.exists(fromPath)
+          const toExists = await fs.exists(toPath)
+
+          expect(toExists && !fromExists).toEqual(true)
+        })
+    )
+  })
+})
+
+
+/* Filesystem */
+
+const emptyFilesystem = async () => {
```

Comment:
Can we move these utilities (they're great!), to a file in `tests/helpers/`?

Same thing with the `pathSegment` and `pathSegmentPair`. Maybe we'll make `privateFileContent` and `publicFileContent`.

(Maybe this is premature abstraction, though, so consider this just an idea)

Reply:
I went ahead and did it! Can totally see that these will see more use than in just these tests.

When I was typing the public decoder, I used some type assertions. Like this one for example:

https://github.com/fission-suite/webnative/blob/3105c2ee8a97dcea9a5c5d52f0e8f1578cda372e/tests/helpers/fileContent.ts#L81

Does this seem alright? I wasn't able to use `BufferSource` in the type signature of the decoder, because we end up with a type mismatch in the tests. The problem is that `fs.read` returns `FileContent` that we decode, but the `TextDecoder` expects a `BufferSource`.

Reply:
Hmm. Yeah, technically the public side will only ever return `Uint8Array` (which is one of the types in the `FileContent` union), so at the moment it's not precise enough. I want to change that for a webnative API redesign in the future, but casting is just fine for now :+1: